File: Den.Dev.Grunt/Den.Dev.Grunt/Core/Modules/SkillModule.cs
// <copyright file="SkillModule.cs" company="Den Delimarsky">
// Developed by Den Delimarsky.
// Den Delimarsky licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
// The underlying API powering Den.Dev.Grunt is managed by Halo Studios and Microsoft. This wrapper is not endorsed by Halo Studios or Microsoft.
// </copyright>

using System.Collections.Generic;
using System.Threading.Tasks;
using Den.Dev.Grunt.Core.Foundation;
using Den.Dev.Grunt.Endpoints;
using Den.Dev.Grunt.Models;
using Den.Dev.Grunt.Models.HaloInfinite;

namespace Den.Dev.Grunt.Core.Modules
{
    /// <summary>
    /// Module for skill-related API operations including CSR and match skill information.
    /// </summary>
    public class SkillModule : ModuleBase
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="SkillModule"/> class.
        /// </summary>
        /// <param name="client">The client instance to use for API requests.</param>
        internal SkillModule(ClientBase client)
            : base(client, HaloCoreEndpoints.SkillOrigin)
        {
        }

        /// <summary>
        /// Returns individual player stats for a given match.
        /// </summary>
        /// <remarks>
        /// Method supports returning results in XML behind the scenes. Class names map to XML data model.
        /// </remarks>
        /// <include file='../../APIDocsExamples/HaloInfinite/Skill_GetMatchPlayerResult.xml' path='example'/>
        /// <param name="matchId">The unique match ID.</param>
        /// <param name="playerIds">Array of player IDs. Each ID string should be in the format of "xuid(XUID_VALUE)".</param>
        /// <returns>An instance of <see cref="MatchSkillInfo"/> representing player skills if the request was successful. Otherwise, returns null.</returns>
        public async Task<HaloApiResultContainer<MatchSkillInfo, RawResponseContainer>> GetMatchPlayerResult(string matchId, List<string> playerIds)
        {
            var formattedPlayerList = string.Join(",", playerIds);
            return await this.GetAsync<MatchSkillInfo>(
                $"/hi/matches/{matchId}/skill?players={formattedPlayerList}",
                useClearance: true);
        }

        /// <summary>
        /// Gets playlist Competitive Skill Rank (CSR) for a player or a set of players.
        /// </summary>
        /// <include file='../../APIDocsExamples/HaloInfinite/Skill_GetPlaylistCsr.xml' path='example'/>
        /// <param name="playlistId">Unique ID for the playlist.</param>
        /// <param name="playerIds">Array of player IDs. Each ID string should be in the format of "xuid(XUID_VALUE)".</param>
        /// <param name="seasonId">Season identifier. Example value is "CsrSeason2-3".</param>
        /// <returns>If successful, an instance of <see cref="PlaylistCsrResultContainer"/> representing player CSRs. Otherwise, returns null.</returns>
        public async Task<HaloApiResultContainer<PlaylistCsrResultContainer, RawResponseContainer>> GetPlaylistCsr(string playlistId, List<string> playerIds, string seasonId = "")
        {
            var formattedPlayerList = string.Join(",", playerIds);
            return await this.GetAsync<PlaylistCsrResultContainer>(
                $"/hi/playlist/{playlistId}/csrs?players={formattedPlayerList}&season={seasonId}",
                useClearance: true);
        }
    }
}

File: src/node/src/modules/halo-infinite/skill.module.ts
import { ModuleBase } from '../base/module-base';
import type { ClientBase } from '../../clients/base/client-base';
import type { HaloApiResult } from '../../models/common/api-result';
import { HALO_CORE_ENDPOINTS } from '../../endpoints/halo-core-endpoints';
import type { MatchSkillInfo, PlaylistCsrResultContainer } from '../../models/halo-infinite/skill';

/**
 * Skill module for CSR (Competitive Skill Rank) queries.
 *
 * Provides access to:
 * - Match skill results (CSR changes after a match)
 * - Playlist CSR for players
 *
 * @example
 * ```typescript
 * // Get CSR for a player in a playlist
 * const csr = await client.skill.getPlaylistCsr('playlist-id', ['xuid1', 'xuid2']);
 *
 * // Get skill results for a match
 * const matchSkill = await client.skill.getMatchPlayerResult('match-id', ['xuid1']);
 * ```
 */
export class SkillModule extends ModuleBase {
  constructor(client: ClientBase) {
    super(client, HALO_CORE_ENDPOINTS.SKILL_ORIGIN);
  }

  /**
   * Get skill/CSR changes for players in a specific match.
   *
   * @param matchId - Match ID in GUID format
   * @param playerIds - List of player XUIDs to query
   * @returns Skill info for each player
   */
  getMatchPlayerResult(
    matchId: string,
    playerIds: string[]
  ): Promise<HaloApiResult<MatchSkillInfo>> {
    this.assertNotEmpty(matchId, 'matchId');
    if (!playerIds.length) {
      throw new Error('playerIds cannot be empty');
    }

    const players = playerIds.map((id) => `xuid(${id})`).join(',');
    return this.get<MatchSkillInfo>(
      `/hi/matches/${matchId}/skill?players=${players}`,
      { useClearance: true }
    );
  }

  /**
   * Get current CSR for players in a specific playlist.
   *
   * @param playlistId - Playlist ID in GUID format
   * @param playerIds - List of player XUIDs to query
   * @param seasonId - Optional season ID for season-specific CSR
   * @returns CSR results for each player
   */
  getPlaylistCsr(
    playlistId: string,
    playerIds: string[],
    seasonId?: string
  ): Promise<HaloApiResult<PlaylistCsrResultContainer>> {
    this.assertNotEmpty(playlistId, 'playlistId');
    if (!playerIds.length) {
      throw new Error('playerIds cannot be empty');
    }

    const players = playerIds.map((id) => `xuid(${id})`).join(',');
    const seasonParam = seasonId ? `&seasonId=${seasonId}` : '';
    return this.get<PlaylistCsrResultContainer>(
      `/hi/playlist/${playlistId}/csrs?players=${players}${seasonParam}`,
      { useClearance: true }
    );
  }
}
